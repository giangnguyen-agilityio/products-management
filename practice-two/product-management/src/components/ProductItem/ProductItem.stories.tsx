import type { Meta, Story } from '@storybook/react'
import ProductItem from './index'
import { ProductsItemProps } from '@constants'

const meta = {
  title: 'Practice Two/ProductItem Component',
  component: ProductItem,
  decorators: [
    Story => (
      <div
        style={{
          backgroundColor: '#532EE7',
          padding: '20px',
        }}
      >
        <Story />
      </div>
    ),
  ],
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
} as Meta

export default meta

const Template: Story<{ product: ProductsItemProps }> = args => (
  <ProductItem {...args} />
)

const exampleProduct: ProductsItemProps = {
  id: 'ID01',
  name: 'iPhone 14 Pro Max',
  image: 'https://assets.nicepagecdn.com/d2cc3eaa/3159880/images/yttyy.jpg',
}

export const Primary = Template.bind({})
Primary.args = {
  product: exampleProduct,
}

const exampleProductWithoutImage: ProductsItemProps = {
  id: 'ID01',
  name: 'iPhone 14 Pro Max',
  image: '',
}

export const ProductItemWithoutImage = Template.bind({})
ProductItemWithoutImage.args = {
  product: exampleProductWithoutImage,
}
