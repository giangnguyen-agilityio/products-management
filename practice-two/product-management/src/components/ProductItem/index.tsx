// Libraries
import React, { memo } from 'react'
import { Link as RouterLink } from 'react-router-dom'
import { GridItem, Box, Image, Text, Button } from '@chakra-ui/react'

// Assets
import imageNotAvailable from '@assets/images/Image_not_available.webp'

// Types
import { IProductItem } from '@types'

interface ProductItemProps {
  product: IProductItem
}

const ProductItem: React.FC<ProductItemProps> = ({ product }) => (
  <GridItem
    className="product-item"
    margin="0 auto"
    key={product.id}
    textAlign="center"
    as="li"
    listStyleType="none"
    borderRadius="large"
    boxShadow="tertiary"
    bgColor="background"
    width="full"
    height={{ base: '516px', md: '490px', lg: '340px', xl: '390px' }}
    padding={{ base: '30px 0', sm: '30px 20px' }}
    position="relative"
  >
    {/* Image Container */}
    <Box className="image-container">
      <Image
        src={product.image || imageNotAvailable}
        alt={product.name}
        width="full"
        height={{
          base: '300px',
          md: '305px',
          lg: '173px',
          xl: '220px',
        }}
        margin="0 auto"
        objectFit="contain"
        loading="lazy"
        transition="transform 0.3s ease-in-out"
        _hover={{
          transform: 'scale(1.05)',
        }}
      />
    </Box>

    {/* Product Name */}
    <Text
      fontWeight="semibold"
      padding="15px 0 0"
      fontSize={{ base: 'extraLarge', md: 'large', lg: 'regular' }}
      color="primary"
      noOfLines={2}
      width={{ xs: '300px', sm: 'full' }}
      margin="0 auto"
      variant="fontTertiary"
    >
      {product.name}
    </Text>
    {/* Detail Link */}
    <Button
      as={RouterLink}
      to={`/products/${product.id}`}
      position="absolute"
      bottom={0}
      left="50%"
      transform="translate(-50%, -50%)"
      fontSize="tiny"
      fontFamily="RobotoCondensed-Bold"
      textTransform="uppercase"
      rounded="full"
      bgColor="primary"
      color="textSecondary"
      width="100px"
      _hover={{
        border: '2px solid',
        color: 'primary',
        bgColor: 'background',
      }}
    >
      detail
    </Button>
  </GridItem>
)

export default memo(ProductItem)
