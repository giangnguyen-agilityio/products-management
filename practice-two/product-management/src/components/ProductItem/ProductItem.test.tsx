import { render } from '@testing-library/react'
import ProductItem from '.'

jest.mock('@assets/images/Image_not_available.webp', () => ({
  default: 'imageNotAvailable',
}))

const mockProduct = {
  id: 'ID01',
  name: 'Sample Product',
  image: 'sample-image.jpg',
}

describe('ProductItem Component', () => {
  it('renders product name correctly', () => {
    const { getByText } = render(<ProductItem product={mockProduct} />)
    const productName = getByText('Sample Product')
    expect(productName).toBeInTheDocument()
  })

  it('renders default image if product image is not available', () => {
    const { container } = render(
      <ProductItem product={{ ...mockProduct, image: '' }} />
    )
    expect(container).toMatchSnapshot()
  })

  it('renders link with correct text', () => {
    const { getByText } = render(<ProductItem product={mockProduct} />)
    const detailLink = getByText('detail')
    expect(detailLink).toBeInTheDocument()
  })

  it('matches snapshot', () => {
    const { container } = render(<ProductItem product={mockProduct} />)
    expect(container).toMatchSnapshot()
  })
})
