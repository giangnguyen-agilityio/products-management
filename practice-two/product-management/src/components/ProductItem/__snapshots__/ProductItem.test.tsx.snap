// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductItem Component matches the snapshot 1`] = `
<div>
  <li
    class="product-item css-gon1ka"
  >
    <div
      class="product-item-content css-55cmdl"
    >
      <div
        class="image-container css-0"
      >
        <img
          alt="Sample Product"
          class="chakra-image css-1goliy0"
          loading="lazy"
          src="sample-image.jpg"
        />
      </div>
      <p
        class="chakra-text css-c8mxau"
      >
        Sample Product
      </p>
      <a
        class="chakra-link css-204pox"
        href="/products/P01"
      >
        detail
      </a>
    </div>
  </li>
</div>
`;

exports[`ProductItem Component renders imageNotAvailable when product image is not available 1`] = `
<div>
  <li
    class="product-item css-gon1ka"
  >
    <div
      class="product-item-content css-55cmdl"
    >
      <div
        class="image-container css-0"
      >
        <img
          alt="Sample Product"
          class="chakra-image css-1goliy0"
          loading="lazy"
          src="[object Object]"
        />
      </div>
      <p
        class="chakra-text css-c8mxau"
      >
        Sample Product
      </p>
      <a
        class="chakra-link css-204pox"
        href="/products/P01"
      >
        detail
      </a>
    </div>
  </li>
  ,
</div>
`;
